core/src/main/java/ai/timefold/solver/core/impl/score/stream/bavet/common/index/IndexerFactory.java

```diff
@@ -526,87 +467,119 @@ public Function<Right_, ai.timefold.solver.core.impl.score.stream.bavet.common.i
         var descendingJoinerTypeMap = joinerTypeMap.descendingMap();
-        Supplier<ai.timefold.solver.core.impl.score.stream.bavet.common.index.Indexer<T>> downstreamIndexerSupplier =
-                NoneIndexer::new;
+        Supplier<Indexer<T>> downstreamIndexerSupplier = NoneIndexer::new;
         var indexPropertyId = descendingJoinerTypeMap.size() - 1;
         for (var entry : descendingJoinerTypeMap.entrySet()) {
             var joinerType = entry.getValue();
             var actualDownstreamIndexerSupplier = downstreamIndexerSupplier;
             var effectivelyFinalIndexPropertyId = indexPropertyId;
             if (joinerType == JoinerType.EQUAL) {
                 downstreamIndexerSupplier =
-                        () -> new ai.timefold.solver.core.impl.score.stream.bavet.common.index.EqualsIndexer<>(
-                                effectivelyFinalIndexPropertyId, actualDownstreamIndexerSupplier);
+                        () -> new EqualsIndexer<>(effectivelyFinalIndexPropertyId, actualDownstreamIndexerSupplier);
             } else {
                 var actualJoinerType = isLeftBridge ? joinerType : joinerType.flip();
-                downstreamIndexerSupplier =
-                        () -> new ai.timefold.solver.core.impl.score.stream.bavet.common.index.ComparisonIndexer<>(
-                                actualJoinerType, effectivelyFinalIndexPropertyId,
-                                actualDownstreamIndexerSupplier);
+                downstreamIndexerSupplier = () -> new ComparisonIndexer<>(actualJoinerType, effectivelyFinalIndexPropertyId,
+                        actualDownstreamIndexerSupplier);
             }
             indexPropertyId--;
         }
         return downstreamIndexerSupplier.get();
     }
 
+    private static <A> UniMapping<A> wrap(Function<A, Object> mapping) {
+        return a -> {
+            var result = mapping.apply(a);
+            return Objects.requireNonNullElse(result, NullIndexProperties.INSTANCE);
+        };
+    }
+
+    private static <A, B> BiMapping<A, B> wrap(BiFunction<A, B, Object> mapping) {
+        return (a, b) -> {
+            var result = mapping.apply(a, b);
+            return Objects.requireNonNullElse(result, NullIndexProperties.INSTANCE);
+        };
+    }
+
+    private static <A, B, C> TriMapping<A, B, C> wrap(TriFunction<A, B, C, Object> mapping) {
+        return (a, b, c) -> {
+            var result = mapping.apply(a, b, c);
+            return Objects.requireNonNullElse(result, NullIndexProperties.INSTANCE);
+        };
+    }
+
+    private static <A, B, C, D> QuadMapping<A, B, C, D> wrap(QuadFunction<A, B, C, D, Object> mapping) {
+        return (a, b, c, d) -> {
+            var result = mapping.apply(a, b, c, d);
+            return Objects.requireNonNullElse(result, NullIndexProperties.INSTANCE);
+        };
+    }
+
+    @FunctionalInterface
+    public interface UniMapping<A> extends Function<A, Object> {
```

Comment:
The logic seems correct. I wonder if the result type could be something other than `Object`, which would clarify the contract for inner nodes.

Reply:
That is unfortunately the one downside of this change - it can only be `Object`.

Previously, it was `IndexProperties`, because everything got mapped to a subtype of `IndexProperties`. But this is no longer the case.

If we have a single key, that key will be returned directly, as opposed to being wrapped in `SingleIndexProperties`. That is where the savings come from. So, unfortunately, this can only ever be `Object` - and that object will sometimes be `IndexProperties`, and sometimes just an arbitrary user type.

I tried to make the situation easier by using `UniMapping` instead of `Function<A, Object>`. This way, when you see `UniMapping`, you already know what the `Object` means, and why.

Reply:
Okay. Saving about `30%` of allocation is a great improvement.